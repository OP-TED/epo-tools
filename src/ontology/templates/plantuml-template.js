import { INHERITANCE, RELATIONSHIP } from '../const.js'

function toPlantuml ({ nodes, edges }) {

  const classDefinitions = nodes.map(node => {
    return nodeTemplate(node, edges.filter(edge => edge.source === node.name).
      filter(edge => edge.type !== INHERITANCE))
  })

  const relations = [
    ...edges.filter(x => x.type === INHERITANCE).map(subclassTemplate),
    ...edges.filter(x => x.type === RELATIONSHIP).map(relationTemplate)]

  return `
@startuml
${classDefinitions.join('\n')}
${noDuplicates(relations).join('\n')}
@enduml
`
}

const nodeTemplate = ({ name }, edges) => `class "${name}" {
${noDuplicates(edges.map(
  x => `  ${x.predicate} : ${x.target} ${displayQuantifiers(x.quantifiers,
    x => `[${x}]`)}`)).join('\n')}
}`

const subclassTemplate = ({
  source, predicate, target,
}) => `"${target}" <|-- "${source}"`

const relationTemplate = ({
  source, predicate, target, quantifiers,
}) => `"${source}" --> ${displayQuantifiers(quantifiers,
  x => `"${x}"`)} "${target}" ${displayPredicateName(predicate)}`

function displayQuantifiers ({ quantifiersDeclared, min, max, raw }, wrap) {
  if (quantifiersDeclared) {
    if (raw) { // prefer raw if declared
      return wrap(`${raw}`)
    } else if (min && max) {
      return wrap(`${min}..${max}`)
    } else if (min) {
      return wrap(`${min}..*`)
    } else if (max) {
      return wrap(`0..${max}`)
    }
  }
  return ''
}

const displayPredicateName = (predicate) => predicate ? `: ${predicate}` : ''

const noDuplicates = (arr) => [...new Set(arr)]

export { toPlantuml }
