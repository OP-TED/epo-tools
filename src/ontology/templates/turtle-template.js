import { ATTRIBUTE, INHERITANCE, RELATIONSHIP } from '../const.js'

const nodeTemplate = ({ name, description }) => `
      # ${name}
      #  ${skosDefinitionTemplate(description)}
      #  a owl:Class .
        `

const literalTemplate = ({
  source, predicate, target, description, quantifiers,
}) => `
    # ${predicate} a owl:DatatypeProperty ;
    #     ${skosDefinitionTemplate(description)}
    #     rdfs:domain  ${source} ;
    #     rdfs:range  ${target} .

    ${shapeIRI({ source, predicate, target })} a sh:NodeShape ;
      sh:targetClass ${source} ;
      sh:property [
        sh:path ${predicate} ;
        sh:datatype ${target} ;
        ${quantifiersTemplate(quantifiers)}
      ] .
    `

const objectTemplate = ({
  source, predicate, target, description, quantifiers,
}) => `
    # ${predicate} a owl:ObjectProperty ;
    #     ${skosDefinitionTemplate(description)}
    #     rdfs:domain  ${source} ;
    #     rdfs:range  ${target} .

    ${shapeIRI({ source, predicate, target })} a sh:NodeShape ;
      sh:targetClass ${source} ;
      sh:property [
        sh:path ${predicate} ;
        sh:nodeKind sh:IRI ;
        ${quantifiersTemplate(quantifiers)}
         sh:targetClass ${target} ;
      ] .
    `

const subclassTemplate = ({ source, predicate, target }) => `
    ${source} rdfs:subClassOf ${target} .
    `

function toTurtle ({ nodes, edges }) {
  const classDefinitions = nodes.map(nodeTemplate)

  const relations = edges.map(edge => {

    if (edge.type === INHERITANCE) {
      return subclassTemplate(edge)
    } else if (edge.type === ATTRIBUTE) {
      return literalTemplate(edge)
    } else if (edge.type === RELATIONSHIP) {
      return objectTemplate(edge)
    } else {
      throw Error(`type '${edge.type}' not implemented`)
    }

  })

  return `
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix schema: <http://schema.org/> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix owl: <http://www.w3.org/2002/07/owl#> .
@prefix skos: <http://www.w3.org/2004/02/skos/core#> .
@prefix dcterms: <http://purl.org/dc/terms/> .
@prefix dct: <http://purl.org/dc/terms/> . # Armonize this!
@prefix foaf: <http://xmlns.com/foaf/0.1/> .

@prefix epo: <http://data.europa.eu/a4g/ontology#> .
@prefix epo-ord: <http://data.europa.eu/a4g/ontology#> .
@prefix epo-ful: <http://data.europa.eu/a4g/ontology#> .
@prefix epo-con: <http://data.europa.eu/a4g/ontology#> .
@prefix epo-cat: <http://data.europa.eu/a4g/ontology#> .
@prefix epo-not: <http://data.europa.eu/a4g/ontology#> .
@prefix epo-acc: <http://data.europa.eu/a4g/ontology#> .
@prefix epo-sub: <http://data.europa.eu/a4g/ontology#> .
@prefix locn: <http://www.w3.org/ns/locn#> .
@prefix cccev: <http://data.europa.eu/m8g/> .
@prefix cv: <http://data.europa.eu/m8g/> . # Armonize this!
@prefix time: <http://www.w3.org/2006/time#> .
@prefix org: <http://www.w3.org/ns/org#> .
@prefix eli: <http://data.europa.eu/eli/ontology#> .
@prefix person: <http://www.w3.org/ns/person#> .
@prefix adms: <http://www.w3.org/ns/adms#> .
@prefix a4g_shape: <http://data.europa.eu/a4g/data-shape#> .

@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix a4g_shape: <http://data.europa.eu/a4g/data-shape#> .
@prefix cpov: <http://data.europa.eu/m8g/cpov> . # Is this true?
@prefix at-voc: <http://publications.europa.eu/resource/authority/> . # Authority tables
@prefix at-voc-new: <http://unknown/> . # Authority tables


  ${[...classDefinitions, ...relations].join('\n')}
`
}

function shapeIRI ({ source, predicate, target }) {
  return `${source}Shape`
}

const quantifiersTemplate = ({
  min, max, quantifiersDeclared,
}) => `
${quantifiersDeclared && max ? `sh:minCount ${min} ;` : ''}
${quantifiersDeclared && max ? `sh:maxCount ${max} ;` : ''}
`

function skosDefinitionTemplate (def) {
  // return def ? `skos:definition """${def}"""" ;` : ''
  return def ? `skos:definition "${(def.split(/\s+/).
    map(x => x.trim()).
    join(' ')).replaceAll('"', '')}" ;` : ''
}

export { toTurtle }
