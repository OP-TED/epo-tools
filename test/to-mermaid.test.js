import { expect } from 'expect'
import toMatchSnapshot from 'expect-mocha-snapshot'
import { describe, it } from 'mocha'
import { battery } from './templates.js'

expect.extend({ toMatchSnapshot })

describe('module', () => {
  for (const [testName, [turtle, term]] of Object.entries(battery)) {
    it(`Test ${testName}`, function () {
      expect([testName, turtle, term]).toMatchSnapshot(this)
    })
  }
})
