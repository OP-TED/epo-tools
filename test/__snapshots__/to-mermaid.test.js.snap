// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`module Test simple 1`] = `
Array [
  "simple",
  "
<a> <b> <c> .
",
  NamedNode {
    "value": "a",
  },
]
`;
